### PR TITLE
Prepare Glow Function for training, replacing constants with placehol…

### PR DESCRIPTION
**Pull request by putivsky**
…ders.

Summary:
Created a comprehensive test for TrainingPreparation utility using MNIST model as an example
Documentation:
Test Plan:
Run ./bin/mnist for both tests.
Please see a detailed explanation of how to fill out the fields in the relevant sections in PULL_REQUEST.md.

### CODE REVIEW

File: lib/Optimizer/TrainingPreparation.cpp [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+      const bool isTrainable =
+          !isSoftMax &&
+          (!isBatchNormalization || (BatchNormalizationNode::MeanIdx != idx &&
+                                     BatchNormalizationNode::VarIdx != idx));
+
+      auto *PH = F->getParent()->createPlaceholder(C->getType(), C->getName(),
+                                                   isTrainable);
+
+      if (isSoftMax) {
+        selected = PH;
```
Comment (jfix71) [inline lib/Optimizer/TrainingPreparation.cpp]: It's possible for there to be multiple SoftMax in a single Function, for example in the inception models where there are intermediate SoftMaxes. I think it makes more sense to remove this finding of the selected PH and just have the caller do it (just with a simple loop over the Function's Nodes) since it understands the design of the model better.

**Comment by putivsky** [comment]
Yes, I have checked the results of training, see line 280 at mnist.cpp
About, BatchNorm - I'm planning (after landing this PR) go back to ResNet20 (includes BN) PR and check training.

File: lib/Optimizer/TrainingPreparation.cpp (hunk 1 above)
Comment (putivsky) [reply lib/Optimizer/TrainingPreparation.cpp]: The general problem even with this PR is that we assuming caller will have enough information about model topology. In practice, caller needs a generic API, which works out of the box for any model. The proposed API can be part of the other framework and should work as it is without additional work. I'm leaning to keep it. Anyway of "selected" is not found it will be nullptr and caller can take actions.

Comment (jfix71) [reply lib/Optimizer/TrainingPreparation.cpp]: The caller has to be aware at least of the inputs it wants to use for the model and associated names. We could always programmatically search through the Function for the corresponding input PHs based on their name to set their associated input Tensors. This is essentially what's going on when you do `loader.getNodeValueByName(inputName)`. You're just doing it differently for the SoftMax here because you know the model has a single SoftMax.

I think it's fine for now in this PR, but I think we should at least add a check here that selected is a nullptr before setting it, and if that fails say the pass only currently supports models with single SoftMax.
